src/terminal/adapter/MacroBuffer.cpp

```diff
@@ -0,0 +1,271 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+#include "precomp.h"
+#include "MacroBuffer.hpp"
+#include "../parser/ascii.hpp"
+#include "../parser/stateMachine.hpp"
+
+using namespace Microsoft::Console::VirtualTerminal;
+
+size_t MacroBuffer::GetSpaceAvailable() const noexcept
+{
+    return MAX_SPACE - _spaceUsed;
+}
+
+uint16_t MacroBuffer::CalculateChecksum() const noexcept
+{
+    // The algorithm that we're using here is intended to match the checksums
+    // produced by the original DEC VT420 terminal. Although note that a real
+    // VT420 would have included the entire macro memory area in the checksum,
+    // which could still contain remnants of previous macro definitions that
+    // are no longer active. We don't replicate that behavior, since that's of
+    // no benefit to applications that might want to use the checksum.

```

Comment:
Doesn't that mean that the checksum won't match the output expected by such applications anyways then?

Reply:
What I'm saying is that nobody is likely to depend on that behavior because it's nondeterministic. As long as they do a full delete when uploading their macros, we should match the expected checksum. And if they aren't doing a full delete, they shouldn't be using a checksum, because the results would be unpredictable. If I'm wrong about that, we can always revisit this implementation.